sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala

```diff
@@ -504,16 +504,29 @@ case class WholeStageCodegenRDD(@transient sc: SparkContext, var source: CodeAnd
       private[this] var iter = computeInternal(split, context)
 
       override def hasNext: Boolean = try {
-        iter.hasNext
-      } catch {
-        case _: ClassCastException =>
-          logInfo(s"ClassCastException, hence recompiling")
-          CodeGenerator.invalidate(source)
-          iter = computeInternal(split, context)
+        try {
           iter.hasNext
+        } catch {
+          case _: ClassCastException =>
+            logInfo(s"ClassCastException, hence recompiling")
+            CodeGenerator.invalidate(source)
+            iter = computeInternal(split, context)
+            iter.hasNext
+        }
+      } catch {
+        case e: Throwable =>
+          logError(s"\n${CodeFormatter.format(source)}")
```

Comment:
Can also add throwable at the start of message just to tell which exception caused the source code. Ideally we should have a separate log-file because if for some reason this happens repeatedly (e.g. a long duration issue like network, or sometimes expected constraint violation exceptions which user intends to ignore) then this can keep of dumping similar code all the time. Though that can be done later, I would be more comfortable if this is at least under some flag for now. We have seen users ending up with tens/hundreds of GBs of logs with the index dump code under assertion failures which was supposed to be very rare.

Reply:
The exception is anyway printed by the task code so wanted to avoid the same exception stack trace to be printed twice. It is easy to link the exception with the code using the thread id. 
As mentioned in a mail to you, ideally we should be dumping the generated code cache in a separate file just like janino that can be used for debugging later. So yeah I agree with you this code should go in a separate file going forward and it can be linked with the exception using an id. 
I will add a system property but that would take effect only when the server is rebooted.

Reply:
@hbhanawat Please read the comment again. Not saying print the stack, just the exception toString to correlate. With concurrent operations it will be difficult to correlate. Can you add that quickly in both places since this has already been merged?
```
          logError(s"\nFailed with exception $e:\n${CodeFormatter.format(source)}")

```

> I will add a system property

Or a GemXD trace flag which can be changed with a system procedure. However, reading that at Spark layer will be a bit long (read "gemfirexd.debug.true" system property, then split on "," if present and search for the flag).